File: Chapter 02/StrategyPatternExample.py
class GiftingStrategy:
    def gift(self, amount):
        pass
class CreditCardDonation(GiftingStrategy):
    def __init__(self, card_number):
        self.card_number = card_number
    def gift(self, amount):
        print(f"Gifted ${amount} via credit card: {self.card_number}")
class CashAppDonation(GiftingStrategy):
    def __init__(self, email):
        self.email = email
    def gift(self, amount):
        print(f"Paid ${amount} via CashApp: {self.email}")
class ShoppingCart:
    def __init__(self):
        self.gifting_strategy = None
    def set_gifting_strategy(self, gifting_strategy):
        self.gifting_strategy = gifting_strategy
    def checkout(self, total_amount):
        self.gifting_strategy.pay(total_amount)
if __name__ == "__main__":
    cart = ShoppingCart()
    credit_card_donation = CreditCardDonation("0000-0000-0000-0000")
    cashApp_donation = CashAppDonation("muzz@email.com")
    cart.set_gifting_strategy(credit_card_donation)
    cart.checkout(319)
    cart.set_gifting_strategy(cashApp_donation)
    cart.checkout(30)

File: Chapter 02/StrategyPatternExample.java
interface GiftingStrategy {
    void gift(int amount);
}

class CreditCardDonation implements GiftingStrategy {
    private String cardNumber;

    public CreditCardDonation(String cardNumber) {
        this.cardNumber = cardNumber;
    }

    public void gift(int amount) {
        System.out.println("Gifted $" + amount + " via credit card: " + cardNumber);
    }
}

class CashAppDonation implements GiftingStrategy {
    private String email;

    public CashAppDonation(String email) {
        this.email = email;
    }

    public void gift(int amount) {
        System.out.println("Gifted $" + amount + " via CashApp: " + email);
    }
}

class ShoppingCart {
    private GiftingStrategy giftStrategy;

    public void setGiftingStrategy(GiftingStrategy giftingStrategy) {
        this.giftingStrategy = giftingStrategy;
    }

    public void checkout(int totalAmount) {
        giftingStrategy.gift(totalAmount);
    }
}


public class StrategyPatternExample {
    public static void main(String[] args) {
        ShoppingCart cart = new ShoppingCart();

        GiftingStrategy creditCardDonation = new CreditCardDonation("0000-0000-0000-0000");
        GiftingStrategy cashAppPayment = new CashAppPayment("muzz@email.com");

        cart.setDonationStrategy(creditCardDonation);
        cart.checkout(319);

        cart.setDonationStrategy(cashAppPayment);
        cart.checkout(30);
    }
}
